pkg/registration/hub/gc/controller.go

```diff
@@ -109,26 +106,33 @@ func NewGCController(
 
 // gc controller is watching cluster and to do these jobs:
 //  1. add a cleanup finalizer to managedCluster if the cluster is not deleting.
-//  2. clean up all rbac and resources in the cluster ns after the cluster is deleted.
+//  2. clean up all rolebinding and resources in the cluster ns after the cluster is deleted.
 func (r *GCController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
 	clusterName := controllerContext.QueueKey()
 	if clusterName == "" || clusterName == factory.DefaultQueueKey {
 		return nil
 	}
 
-	originalCluster, err := r.clusterLister.Get(clusterName)
-	switch {
-	case errors.IsNotFound(err):
-		return nil
-	case err != nil:
+	// cluster could be nil, that means the cluster is gone but the gc is not finished.
+	cluster, err := r.clusterLister.Get(clusterName)
+	if err != nil && !apierrors.IsNotFound(err) {
 		return err
 	}
 
-	cluster := originalCluster.DeepCopy()
+	var copyCluster *clusterv1.ManagedCluster
+	if cluster != nil {
+		if cluster.DeletionTimestamp.IsZero() {
+			_, err = r.clusterPatcher.AddFinalizer(ctx, cluster, clusterv1.ManagedClusterFinalizer)
```

Comment:
with this finalizer, why is it possible the cluster was deleted before resource cleanup?

Reply:
There is a case that will remove the finalizer after delete the rbac when gc feature gate is disable, but the finalizer of work rolebinding can only be removed after all works are deleted. in this case the cluster may be gone.